File: java/src/main/java/com/example/RandomUtil.java
package com.example;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class RandomUtil {
    private static final boolean ALL_ASCII = false;
    private static final String[] NOT_ASCII = new String[]{"\uD842\uDFB7", "一", "二", "三", "四", "五"};
    private static final char[] SPECIAL_CHARACTERS = "~@#$%^&*(){}[]<>:\"?;',./".toCharArray();
    public static final Random random = new Random(1);
    private static final int ARRAY_SIZE = 15;
    private static final int P = 5;

    public static boolean randomBoolean() {
        return random.nextBoolean();
    }

    public static float randomFloat() {
        int i = random.nextInt(10000);
        if (i % P == 0) {
            return 0f;
        }
        return random.nextFloat() * i;
    }

    public static double randomDouble() {
        int i = random.nextInt(1000000);
        if (i % P == 0) {
            return 0D;
        }
        return random.nextDouble() * random.nextInt(1000000);
    }

    public static double randomSmallDouble() {
        int i = random.nextInt();
        int flag = i % 4;
        if (flag == 0) {
            return 0D;
        }
        if (flag == 1) {
            return i & 0x1F;
        }
        if (flag == 2) {
            return i & 0xFFFFF;
        }
        return (i & 0xFFFFF) / 10.D;
    }

    public static int randomCount(int n) {
        return random.nextInt(n);
    }

    public static int randomInt() {
        int i = random.nextInt();
        if (i % P == 0) {
            return 0;
        }
        int shift = (i & 3) << 3;
        return i >>> shift;
    }

    public static int randomSignedInt() {
        int i = randomInt();
        return (i & 1) == 0 ? i : -i;
    }

    public static long randomLong() {
        long l = random.nextLong();
        if (l % P == 0) {
            return 0;
        }
        int shift = ((int) (l & 7)) << 3;
        return l >>> shift;
    }

    public static long randomSignedLong() {
        long l = randomLong();
        return (l & 1) == 1 ? l : -l;
    }

    public static String randomString() {
        int i = random.nextInt();
        if (i % P == 0) {
            return "";
        }
        return randomStr(i % 300);
    }

    public static String randomNullableString() {
        if ((random.nextInt() & 7) == 0) {
            return null;
        } else {
            return randomString();
        }
    }

    public static String randomShortString() {
        int i = random.nextInt();
        if(i % 10 == 0){
            return randomStr(random.nextInt(100));
        }else{
            return randomStr(random.nextInt(50));
        }
    }

    private static String randomStr(int maxLen) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < maxLen; i++) {
            int a =  random.nextInt(Integer.MAX_VALUE) % 50;
            int r = random.nextInt(Integer.MAX_VALUE);
            if (!ALL_ASCII && a < 4) {
                builder.append(NOT_ASCII[r % NOT_ASCII.length]);
            } else if (a < 10) {
                builder.append(SPECIAL_CHARACTERS[r % SPECIAL_CHARACTERS.length]);
            } else if (a < 20) {
                builder.append((char) ('0' + r % 10));
            } else if (a < 30) {
                builder.append((char) ('A' + r % 26));
            } else {
                builder.append((char) ('a' + r % 26));
            }
        }
        return builder.toString();
    }

    public static List<Boolean> randomBoolList() {
        int n = random.nextInt(ARRAY_SIZE);
        List<Boolean> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            a.add(random.nextBoolean());
        }
        return a;
    }

    public static int[] randomSingedIntArray(int n){
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = randomSignedInt();
        }
        return a;
    }

    public static List<Integer> randomIntList() {
        int n = random.nextInt(ARRAY_SIZE);
        List<Integer> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            a.add(Math.abs(random.nextInt()));
        }
        return a;
    }

    public static long[] randomSignedLongArray(int n){
        long[] a = new long[n];
        for (int i = 0; i < n; i++) {
            a[i] = randomSignedLong();
        }
        return a;
    }

    public static List<Long> randomLongList() {
        int n = random.nextInt(ARRAY_SIZE);
        List<Long> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            a.add(Math.abs(random.nextLong()));
        }
        return a;
    }

    public static List<Float> randomFloatList() {
        int n = random.nextInt(ARRAY_SIZE);
        List<Float> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            a.add(random.nextFloat());
        }
        return a;
    }

    public static double[] randomDoubleArray(int n) {
        double[] a = new double[n];
        for (int i = 0; i < n; i++) {
            a[i] = randomSmallDouble();
        }
        return a;
    }

    public static List<Double> randomDoubleList() {
        int n = random.nextInt(ARRAY_SIZE);
        List<Double> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            a.add(random.nextDouble() * randomInt());
        }
        return a;
    }

    public static List<String> randomStringList() {
        int n = random.nextInt(ARRAY_SIZE);
        String[] a = new String[n];
        for (int i = 0; i < n; i++) {
            a[i] = randomShortString();
        }
        return Arrays.asList(a);
    }

    public static boolean[] toBooleanList(List<Boolean> list) {
        int n = list.size();
        Boolean[] a = new Boolean[n];
        boolean[] b = new boolean[n];
        a = list.toArray(a);
        for (int i = 0; i < n; i++) {
            b[i] = a[i];
        }
        return b;
    }

    public static int[] toIntList(List<Integer> list) {
        int n = list.size();
        Integer[] a = new Integer[n];
        int[] b = new int[n];
        a = list.toArray(a);
        for (int i = 0; i < n; i++) {
            b[i] = a[i];
        }
        return b;
    }

    public static float[] toFloatList(List<Float> list) {
        int n = list.size();
        Float[] a = new Float[n];
        float[] b = new float[n];
        a = list.toArray(a);
        for (int i = 0; i < n; i++) {
            b[i] = a[i];
        }
        return b;
    }

    public static long[] toLongList(List<Long> list) {
        int n = list.size();
        Long[] a = new Long[n];
        long[] b = new long[n];
        a = list.toArray(a);
        for (int i = 0; i < n; i++) {
            b[i] = a[i];
        }
        return b;
    }

    public static double[] toDoubleList(List<Double> list) {
        int n = list.size();
        Double[] a = new Double[n];
        double[] b = new double[n];
        a = list.toArray(a);
        for (int i = 0; i < n; i++) {
            b[i] = a[i];
        }
        return b;
    }

    public static String[] toStringList(List<String> list) {
        int n = list.size();
        String[] a = new String[n];
        a = list.toArray(a);
        return a;
    }
}

File: Packable-Kotlin/src/main/kotlin/example/RandomUtil.kt
package example

import kotlin.math.abs
import kotlin.random.Random


object RandomUtil {
    private const val ALL_ASCII = false
    private val NOT_ASCII = arrayOf("\uD842\uDFB7", "一", "二", "三", "四", "五")
    private val SPECIAL_CHARACTERS = "~@#$%^&*(){}[]<>:\"?;',./".toCharArray()
    private val random = Random(1)
    private const val ARRAY_SIZE = 15
    private const val P = 5
    fun randomBoolean(): Boolean {
        return random.nextBoolean()
    }

    fun randomFloat(): Float {
        return random.nextFloat() * random.nextInt(10000)
    }

    fun randomDouble(): Double {
        return random.nextDouble() * random.nextInt(1000000)
    }

    fun randomSmallDouble(): Double {
        val i = random.nextInt()
        val flag = i % 4
        if (flag == 0) {
            return 0.0
        }
        if (flag == 1) {
            return (i and 0x1F).toDouble()
        }
        return if (flag == 2) {
            (i and 0xFFFFF).toDouble()
        } else (i and 0xFFFFF) / 10.0
    }

    fun randomCount(n: Int): Int {
        return random.nextInt(n)
    }

    fun randomInt(): Int {
        val i = random.nextInt()
        if (i % P == 0) {
            return 0
        }
        val shift = i and 3 shl 3
        return i ushr shift
    }

    fun randomLong(): Long {
        val l = random.nextLong()
        if (l % P == 0L) {
            return 0
        }
        val shift = (l and 7L).toInt() shl 3
        return l ushr shift
    }

    fun randomString(): String {
        val i = random.nextInt()
        return if (i % P == 0) {
            ""
        } else randomStr(i % 300)
    }

    fun randomNullableString(): String? {
        return if (random.nextInt() and 7 == 0) {
            null
        } else {
            randomString()
        }
    }

    fun randomShortString(): String {
        val i = random.nextInt()
        return if (i % 10 == 0) {
            randomStr(random.nextInt(100))
        } else {
            randomStr(random.nextInt(50))
        }
    }

    private fun randomStr(maxLen: Int): String {
        val builder = StringBuilder()
        for (i in 0 until maxLen) {
            val a = random.nextInt(Int.MAX_VALUE) % 50
            val r = random.nextInt(Int.MAX_VALUE)
            if (!ALL_ASCII && a < 4) {
                builder.append(NOT_ASCII[r % NOT_ASCII.size])
            } else if (a < 10) {
                builder.append(SPECIAL_CHARACTERS[r % SPECIAL_CHARACTERS.size])
            } else if (a < 20) {
                builder.append(('0'.code + r % 10).toChar())
            } else if (a < 30) {
                builder.append(('A'.code + r % 26).toChar())
            } else {
                builder.append(('a'.code + r % 26).toChar())
            }
        }
        return builder.toString()
    }

    fun randomBoolList(): BooleanArray {
        val n = random.nextInt(ARRAY_SIZE)
        val a = BooleanArray(n)
        for (i in 0 until n) {
            a[i] = random.nextBoolean()
        }
        return a
    }

    fun randomIntList(): IntArray {
        val n = random.nextInt(ARRAY_SIZE)
        val a = IntArray(n)
        for (i in 0 until n) {
            a[i] = abs(random.nextInt())
        }
        return a
    }

    fun randomLongList(): LongArray {
        val n = random.nextInt(ARRAY_SIZE)
        val a = LongArray(n)
        for (i in 0 until n) {
            a[i] = abs(random.nextLong())
        }
        return a
    }

    fun randomFloatList(): FloatArray {
        val n = random.nextInt(ARRAY_SIZE)
        val a = FloatArray(n)
        for (i in 0 until n) {
            a[i] = random.nextFloat()
        }
        return a
    }

    fun randomDoubleList(): DoubleArray {
        val n = random.nextInt(ARRAY_SIZE)
        val a = DoubleArray(n)
        for (i in 0 until n) {
            a[i] = random.nextDouble() * randomInt()
        }
        return a
    }

    fun randomStringList(): Array<String> {
        val n = random.nextInt(ARRAY_SIZE)
        val a = arrayOfNulls<String>(n)
        for (i in 0 until n) {
            a[i] = randomShortString()
        }
        return a.mapNotNull { it }.toTypedArray()
    }
}

